numpy/core/src/npysort/selection.cpp

```diff
@@ -377,269 +378,105 @@ introselect_(type *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                                             npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_long(npy_long *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                  npy_intp *pivots, npy_intp *npiv, void *NOT_USED)
-{
-    return introselect_<npy::long_tag, true>(v, tosort, num, kth, pivots,
-                                             npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_ulong(npy_ulong *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                   npy_intp *pivots, npy_intp *npiv, void *NOT_USED)
-{
-    return introselect_<npy::ulong_tag, true>(v, tosort, num, kth, pivots,
-                                              npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_longlong(npy_longlong *v, npy_intp *tosort, npy_intp num,
-                      npy_intp kth, npy_intp *pivots, npy_intp *npiv,
-                      void *NOT_USED)
-{
-    return introselect_<npy::longlong_tag, true>(v, tosort, num, kth, pivots,
-                                                 npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_ulonglong(npy_ulonglong *v, npy_intp *tosort, npy_intp num,
-                       npy_intp kth, npy_intp *pivots, npy_intp *npiv,
-                       void *NOT_USED)
-{
-    return introselect_<npy::ulonglong_tag, true>(v, tosort, num, kth, pivots,
-                                                  npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_half(npy_half *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                  npy_intp *pivots, npy_intp *npiv, void *NOT_USED)
-{
-    return introselect_<npy::half_tag, true>(v, tosort, num, kth, pivots,
-                                             npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_float(npy_float *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                   npy_intp *pivots, npy_intp *npiv, void *NOT_USED)
+template <typename Tag>
+static int
+introselect_arg(void *v, npy_intp *tosort, npy_intp num, npy_intp kth,
+                npy_intp *pivots, npy_intp *npiv, void *)
 {
-    return introselect_<npy::float_tag, true>(v, tosort, num, kth, pivots,
-                                              npiv);
+    return introselect_<Tag, true>((typename Tag::type *)v, tosort, num, kth,
+                                   pivots, npiv);
 }
-NPY_NO_EXPORT int
-aintroselect_double(npy_double *v, npy_intp *tosort, npy_intp num,
-                    npy_intp kth, npy_intp *pivots, npy_intp *npiv,
-                    void *NOT_USED)
+
+struct arg_map {
+    int typenum;
+    PyArray_PartitionFunc *part[NPY_NSELECTS];
+    PyArray_ArgPartitionFunc *argpart[NPY_NSELECTS];
+};
+
+template <class... Tags>
+static constexpr std::array<arg_map, sizeof...(Tags)>
+make_partition_map(npy::taglist<Tags...>)
 {
-    return introselect_<npy::double_tag, true>(v, tosort, num, kth, pivots,
-                                               npiv);
+    return std::array<arg_map, sizeof...(Tags)>{
+            arg_map{Tags::type_value, &introselect_noarg<Tags>,
```

Comment:
This is the actual initialization. Each tag leads to the creation of an element of  `arg_map` type